radixdlt-java/src/main/java/com/radixdlt/client/application/translate/tokens/StakeTokensMapper.java

```diff
@@ -0,0 +1,111 @@
+import com.radixdlt.client.application.translate.StatefulActionToParticleGroupsMapper;
+import com.radixdlt.client.atommodel.tokens.StakedTokensParticle;
+import com.radixdlt.client.atommodel.tokens.TransferrableTokensParticle;
+import com.radixdlt.client.core.atoms.ParticleGroup;
+import com.radixdlt.client.core.atoms.particles.Particle;
+import com.radixdlt.client.core.atoms.particles.Spin;
+import com.radixdlt.client.core.fungible.FungibleParticleTransitioner;
+import com.radixdlt.client.core.fungible.FungibleParticleTransitioner.FungibleParticleTransition;
+import com.radixdlt.client.core.fungible.NotEnoughFungiblesException;
+import com.radixdlt.identifiers.RRI;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Maps a stake tokens action to the particles necessary to be included in an atom.
+ */
+public class StakeTokensMapper implements StatefulActionToParticleGroupsMapper<StakeTokensAction> {
+	public StakeTokensMapper() {
+	}
+
+	@Override
+	public Set<ShardedParticleStateId> requiredState(StakeTokensAction action) {
+		return Collections.singleton(ShardedParticleStateId.of(TransferrableTokensParticle.class, action.getFrom()));
+	}
+
+	@Override
+	public List<ParticleGroup> mapToParticleGroups(StakeTokensAction stake, Stream<Particle> store) throws StageActionException {
+		final RRI tokenRef = stake.getRRI();
+
+		List<TransferrableTokensParticle> tokenConsumables = store
+			.map(TransferrableTokensParticle.class::cast)
+			.filter(p -> p.getTokenDefinitionReference().equals(tokenRef))
+			.collect(Collectors.toList());
+
+		final FungibleParticleTransitioner<TransferrableTokensParticle, StakedTokensParticle> transitioner =
+			new FungibleParticleTransitioner<>(
+				(amt, consumable) -> new StakedTokensParticle(
+					stake.getDelegate(),
+					amt,
+					consumable.getGranularity(),
+					stake.getFrom(),
+					System.nanoTime(),
+					consumable.getTokenDefinitionReference(),
+					System.currentTimeMillis() / 60000L + 60000L,
+					consumable.getTokenPermissions()
+				),
+				stakedTokens -> stakedTokens,
+				(amt, consumable) -> new TransferrableTokensParticle(
+					amt,
+					consumable.getGranularity(),
+					consumable.getAddress(),
+					System.nanoTime(),
+					consumable.getTokenDefinitionReference(),
+					System.currentTimeMillis() / 60000L + 60000L,
+					consumable.getTokenPermissions()
+				),
+				tokens -> tokens,
+				TransferrableTokensParticle::getAmount
+			);
+
+
+		try {
+			FungibleParticleTransition<TransferrableTokensParticle, StakedTokensParticle> transition = transitioner.createTransition(
+				tokenConsumables,
+				TokenUnitConversions.unitsToSubunits(stake.getAmount())
+			);
+
+			ParticleGroup.ParticleGroupBuilder particleGroupBuilder = ParticleGroup.builder();
+			transition.getRemoved().stream().map(t -> (Particle) t).forEach(p -> particleGroupBuilder.addParticle(p, Spin.DOWN));
```

Comment:
These would be way more efficient and legible without creating a stream as:
```
transition.getRemoved().forEach(t -> particleGroupBuilder.addParticle((Particle) p, Spin.DOWN));
```

Reply:
True that, good point.